File: src/agent/systemPrompt.ts
/**
 * Builds the dynamic system prompt that prefaces every agent turn.
 *
 * Pure function. Inputs:
 *   - `now`: current Date (so the LLM knows the user's current time)
 *   - `timezone`: IANA timezone string (e.g. 'Asia/Kolkata')
 *   - `preferences`: flat key/value map from preferencesStore.snapshot
 *   - `connectedProviders`: list of currently connected providers
 *
 * Output: the canonical `system` message content per the engineering
 * directive. Deterministic for fixed inputs (locked by unit test).
 */

import { type Provider } from '../types/auth';

/**
 * Light user identity passed into the prompt so the agent can address
 * the user by name or email when natural. The fields are independent
 * (some users connect Google but never share a profile name; some use
 * the OpenAI key path only and have no email at all).
 */
export interface SystemPromptUser {
  readonly email: string | null;
  readonly displayName: string | null;
}

export interface SystemPromptInput {
  readonly now: Date;
  readonly timezone: string;
  readonly preferences: Readonly<Record<string, string>>;
  readonly connectedProviders: readonly Provider[];
  readonly user?: SystemPromptUser;
}

const formatNow = (now: Date, timezone: string): string => {
  // Intl.DateTimeFormat is available in modern Hermes/Node. We pin
  // the long-form weekday + day + month + year + 24h time so the LLM
  // has unambiguous temporal grounding.
  const fmt = new Intl.DateTimeFormat('en-GB', {
    weekday: 'long',
    day: 'numeric',
    month: 'long',
    year: 'numeric',
    hour: '2-digit',
    minute: '2-digit',
    hour12: false,
    timeZone: timezone,
  });
  return fmt.format(now);
};

const formatPreferences = (prefs: Readonly<Record<string, string>>): string => {
  const keys = Object.keys(prefs).sort();
  if (keys.length === 0) return '  (no preferences set)';
  return keys.map((k) => `  - ${k}: ${prefs[k]}`).join('\n');
};

const formatConnections = (providers: readonly Provider[]): string => {
  if (providers.length === 0) return '  (none — ask the user to connect a service in the Vault)';
  const labels: Record<Provider, string> = {
    google: 'Gmail + Google Calendar',
    whatsapp: 'WhatsApp',
    openai: 'OpenAI',
  };
  return providers.map((p) => `  - ${labels[p]}`).join('\n');
};

const BASE = `You are Nexus, a personal AI agent that runs on the user's mobile device. You operate on their digital life through a set of tools — Gmail, Google Calendar, contacts, and (later) WhatsApp. Be professional, helpful, and concise. Never invent information; call tools to get the truth.`;

const RULES = [
  'Rules:',
  '- ALWAYS confirm before sending messages, sending email, or creating events. The runtime will surface a confirmation card to the user — your job is to present the drafted action clearly.',
  '- If a tool returns an error, apologize, explain the failure in plain language, and suggest the next step (e.g. "reconnect Google in the Vault", "try again in a moment", or "check the recipient address").',
  '- When the user mentions a person by name (e.g. "send X to my brother"), call system_contacts_search FIRST to resolve the phone number or email before calling the messaging tool.',
  '- Times are interpreted in the timezone shown above unless the user gives an explicit one.',
];

const formatUser = (user: SystemPromptUser | undefined): string | null => {
  if (!user) return null;
  const parts: string[] = [];
  if (user.displayName !== null && user.displayName.length > 0) {
    parts.push(`Name: ${user.displayName}`);
  }
  if (user.email !== null && user.email.length > 0) {
    parts.push(`Email: ${user.email}`);
  }
  if (parts.length === 0) return null;
  return parts.map((p) => `  - ${p}`).join('\n');
};

export const build = (input: SystemPromptInput): string => {
  const sections: string[] = [
    BASE,
    `Current time: ${formatNow(input.now, input.timezone)} (${input.timezone})`,
  ];
  const userBlock = formatUser(input.user);
  if (userBlock !== null) {
    sections.push('User identity:');
    sections.push(userBlock);
  }
  sections.push('User preferences:');
  sections.push(formatPreferences(input.preferences));
  sections.push('Connected services:');
  sections.push(formatConnections(input.connectedProviders));
  sections.push(RULES.join('\n'));
  return sections.join('\n\n');
};

File: app/src/main/java/com/nexus/app/domain/agent/SystemPrompt.kt
package com.nexus.app.domain.agent

import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

object SystemPromptBuilder {

    private const val PERSONA = """You are Nexus, a private personal AI agent that lives entirely on the user's phone.
You can read email, draft messages, manage calendar events, and search local memories.
You speak briefly, with a calm, helpful, professional tone.
You never invent data: if a tool call fails, explain the failure naturally and offer next steps.
You always follow the confirmation rule below."""

    private const val RULES = """RULES:
- ALWAYS confirm with the user before sending an email or creating a calendar event.
- If a tool returns an error, apologize once and explain the cause clearly.
- If the user asks something that needs a service that is not connected, tell them to connect it in Vault."""

    fun build(
        preferences: Map<String, String>,
        connectedProviders: Set<String>,
        timeZone: TimeZone = TimeZone.getDefault(),
        now: Date = Date()
    ): String {
        val time = SimpleDateFormat("EEEE, d MMMM yyyy, HH:mm zzz", Locale.US).apply {
            this.timeZone = timeZone
        }.format(now)

        val prefsBlock = if (preferences.isEmpty()) "(none)" else preferences.entries.joinToString("\n") { (k, v) ->
            "- $k: $v"
        }
        val servicesBlock = if (connectedProviders.isEmpty()) "(none yet — ask the user to connect Google in Vault if needed)"
        else connectedProviders.joinToString(", ")

        return buildString {
            append(PERSONA)
            append("\n\nCurrent time: $time")
            append("\n\nUser Preferences:\n").append(prefsBlock)
            append("\n\nConnected Services: ").append(servicesBlock)
            append("\n\n").append(RULES)
        }
    }
}
